hal/src/electron/cellular_hal.cpp

```diff
@@ -176,13 +180,93 @@ void cellular_cancel(bool cancel, bool calledFromISR, void*)
+            // LISA-U200-01S / LISA-U200-02S / LISA-U200-52S / LISA-U200-62S / LISA-U230 / LISA-U260 /
+            // LISA-U270 / LISA-U1
+            // The <qual> parameter is not updated in GPRS and EGPRS packet transfer mode
+            if (status.act == ACT_GSM) {
+                // Convert to BER (% * 100), see 3GPP TS 45.008 8.2.4
+                signalext->ber = (status.rxqual != 99) ? berMapping[status.rxqual] : std::numeric_limits<int32_t>::min();
+            } else /* status.act == ACT_EDGE */ {
+                // Convert to MEAN_BEP level first
+                // See u-blox AT Reference Manual:
+                // In 2G RAT EGPRS packet transfer mode indicates the Mean Bit Error Probability (BEP) of a radio
+                // block. 3GPP TS 45.008 [148] specifies the range 0-31 for the Mean BEP which is mapped to
+                // the range 0-7 of <qual>
+                int bepLevel = (status.rxqual != 99) ? (7 - status.rxqual) * 31 / 7 : std::numeric_limits<int32_t>::min();
+                // Convert to log10(MEAN_BEP) multiplied by 100, see 3GPP TS 45.008 10.2.3.3
+                // Uses QPSK table
+                signalext->bep = bepLevel >= 0 ? (-(bepLevel - 31) * 10 - 3.7) : bepLevel;
+            }
+
+            // RSSI in % [0, 100] based on [-111, -48] range mapped to [0, 65535] integer range
+            signalext->strength = (status.rxlev != 99) ? status.rxlev * 65535 / 63 : std::numeric_limits<int32_t>::min();
+            // Quality based on RXQUAL in % [0, 100] mapped to [0, 65535] integer range
+            signalext->quality = (status.rxqual != 99) ? (7 - status.rxqual) * 65535 / 7 : std::numeric_limits<int32_t>::min();
+            break;
+        case ACT_UTRAN:
+            // Convert to dBm [-121, -25], see 3GPP TS 25.133 9.1.1.3
+            // Reported multiplied by 100
+            signalext->rscp = ((status.rscp != 255) ? status.rscp - 116 : -121) * 100;
+            // Convert to Ec/Io (dB) [-24.5, 0], see 3GPP TS 25.133 9.1.2.3
+            // Report multiplied by 100
+            signalext->ecno = (status.ecno != 255) ? status.ecno * 50 - 2450 : -2450;
+
+            // RSCP in % [0, 100] based on [-121, -25] range mapped to [0, 65535] integer range
+            signalext->strength = (status.rscp != 255) ? (status.rscp + 5) * 65535 / 96 : std::numeric_limits<int32_t>::min();
+            // Quality based on Ec/Io in % [0, 100] mapped to [0,65535] integer range
+            signalext->quality = (status.ecno != 255) ? status.ecno * 65535 / 49 : std::numeric_limits<int32_t>::min();
+            break;
+        default:
+            res = SYSTEM_ERROR_UNKNOWN;
+            signalext->rssi = std::numeric_limits<int32_t>::min();
+            signalext->qual = std::numeric_limits<int32_t>::min();
+            signalext->strength = 0;
+            signalext->quality = 0;
+            break;
+        }
+    }
+    return res;
```

Comment:
Perhaps factor out fetching the data from the modem from computing the percentages so that the computation can be more easily tested.